### PR TITLE
CODENVY-413: Make it possible to mark MachineNode as 'sheduled for maintenace'

### PR DESCRIPTION
**Pull request by mmorhun**
This and [this](https://github.com/eclipse/che/pull/1705) PRs add ability to mark node for maintenance (it means that new containers won't start on this node).
To mark node for maintenance just add com.codenvy.node-state=maintenance label to docker on the node.
To return node to normal state just remove the label.

To add label to docker:
1. edit docker config file, add label
2. reload (without restart!) docker config (use SIGHUP signal)
Note, that it can take about a minute or so while swarm "see" the label.
To remove label just erase it from config and reload it again.

### CODE REVIEW

**Comment by mmorhun** [comment]
@garagatyi take a look please

**Comment by garagatyi** [comment]
Please ensure that your commit message is git formatting friendly

File: assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiModule.java [hunk 1]
```diff
@@ -296,6 +296,15 @@ protected void configure() {
 
         bind(org.eclipse.che.plugin.docker.client.DockerConnector.class).to(com.codenvy.swarm.client.SwarmDockerConnector.class);
 
+        Multibinder<String> allMachinesEnvVars = Multibinder.newSetBinder(binder(),
+                                                                          String.class,
+                                                                          Names.named("machine.docker.machine_env"))
+                                                            .permitDuplicates();
+        allMachinesEnvVars.addBinding()
+                          .toProvider(org.eclipse.che.plugin.docker.machine.ext.provider.ApiEndpointEnvVariableProvider.class);
```
Comment (garagatyi) [inline assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiModule.java]: Are you sure that binding of ApiEndpointEnvVariableProvider is needed here?

File: assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiModule.java [hunk 2]
```diff
@@ -296,6 +296,15 @@ protected void configure() {
 
         bind(org.eclipse.che.plugin.docker.client.DockerConnector.class).to(com.codenvy.swarm.client.SwarmDockerConnector.class);
 
+        Multibinder<String> allMachinesEnvVars = Multibinder.newSetBinder(binder(),
+                                                                          String.class,
+                                                                          Names.named("machine.docker.machine_env"))
+                                                            .permitDuplicates();
```
Comment (garagatyi) [inline assembly/onpremises-ide-packaging-war-platform-api/src/main/java/com/codenvy/api/deploy/OnPremisesIdeApiModule.java]: Remove that. It is not needed.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java [hunk 3]
```diff
@@ -121,6 +129,49 @@ protected void pullImage(MachineConfig machineConfig, String machineImageName, P
         MachineConfigImpl machineConfigForBuild = new MachineConfigImpl(machineConfig);
         machineConfigForBuild.setSource(new MachineSourceImpl(DockerInstanceProvider.DOCKER_FILE_TYPE)
                                                 .setContent("FROM " + machineConfig.getSource().getLocation()));
-        super.buildImage(machineConfigForBuild, machineImageName, true, progressMonitor);
+        buildImage(machineConfigForBuild, machineImageName, true, progressMonitor);
+    }
+
+    @Override
+    protected void buildImage(MachineConfig machineConfig,
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java]: Add javadoc with description what special this method does. Don't forget to inherit docs

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java [hunk 4]
```diff
@@ -121,6 +129,49 @@ protected void pullImage(MachineConfig machineConfig, String machineImageName, P
...
+                              String machineImageName,
+                              boolean doForcePullOnBuild,
+                              ProgressMonitor progressMonitor)
+            throws MachineException {
+
+        Recipe recipe = recipeRetriever.getRecipe(machineConfig);
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java]: These components can be set in constructor. Don't use protected scope for them

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java [hunk 5]
```diff
@@ -121,6 +129,49 @@ protected void pullImage(MachineConfig machineConfig, String machineImageName, P
...
+        File workDir = null;
+        try {
+            // build docker image
+            workDir = Files.createTempDirectory(null).toFile();
+            final File dockerfileFile = new File(workDir, "Dockerfile");
+            dockerfile.writeDockerfile(dockerfileFile);
+
+            docker.buildImage(BuildImageParams.create(dockerfileFile)
+                                              .withRemoveIntermediateContainers(true)
+                                              .withRemoveIntermediateContainersWithForce(true)
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedDockerInstanceProvider.java]: As docs to this method says it includes regular rm.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/MaintenanceConstraintProvider.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Codenvy S.A..
+ */
+package com.codenvy.machine;
+
+import com.google.inject.Provider;
+
+/**
+ * @author Mykola Morhun
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/MaintenanceConstraintProvider.java]: Description?

**Comment by garagatyi** [comment]
LGTM

**Comment by mmorhun** [comment]
@sleshchenko take a look please

**Comment by sleshchenko** [comment]
OK for me

**Pull request merged**